File: src/intelliticket_backend/schemas/tickets.py
from __future__ import annotations

from datetime import UTC, datetime
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator


class DataMode(StrEnum):
    MOCK = "mock"
    DEMO = "demo"
    REAL = "real"


class DeskId(StrEnum):
    OPS = "ops"
    SUPPORT = "support"


class TicketCategory(StrEnum):
    OPS_ALERT = "ops_alert"
    SUPPORT_REQUEST = "support_request"


class TicketPriority(StrEnum):
    P1 = "P1"
    P2 = "P2"
    P3 = "P3"
    P4 = "P4"


class EvidenceRef(BaseModel):
    evidence_id: str


class Evidence(BaseModel):
    evidence_id: str
    source_type: str
    source_id: str
    source_name: str
    observed_at: str | None = None
    retrieved_at: str | None = None
    service: str | None = None
    metric_name: str | None = None
    value: Any | None = None
    unit: str | None = None
    quality: str
    data_mode: DataMode
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    summary: str
    trace_uri: str | None = None
    freshness: str | None = None
    quality_reason: str | None = None
    producer: str | None = None
    run_id: str | None = None


class TicketProcessRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=2000)
    data_mode: DataMode = DataMode.MOCK
    desk_id: DeskId = DeskId.OPS
    submitter: str | None = None
    operator_id: str | None = None

    @field_validator("text")
    @classmethod
    def strip_text(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("工单内容不能为空")
        return stripped


class TicketSubmitRequest(BaseModel):
    """员工提交工单（只创建，不处理）。提交人由认证 token 确定。"""

    text: str = Field(..., min_length=1, max_length=2000)
    data_mode: DataMode = DataMode.MOCK
    desk_id: DeskId = DeskId.OPS

    @field_validator("text")
    @classmethod
    def strip_text(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("工单内容不能为空")
        return stripped


class TicketSubmitResponse(BaseModel):
    """工单提交结果。"""

    ticket_id: str
    status: str  # "pending"
    created_at: str
    text: str
    desk_id: DeskId
    submitter: str


class TicketClassification(BaseModel):
    category: TicketCategory
    summary: str
    affected_service: str | None
    symptoms: list[str]
    priority: TicketPriority
    priority_reason: str
    extracted_metrics: dict[str, Any] = Field(default_factory=dict)
    evidence_ids: list[str] = Field(default_factory=list)


class ServiceContext(BaseModel):
    service_id: str
    name: str
    display_name: str
    aliases: list[str]
    owner_team: str
    criticality: str
    dependencies: list[str] = Field(default_factory=list)
    data_mode: DataMode


class MetricSnapshot(BaseModel):
    evidence_id: str
    metric_name: str
    value: Any
    unit: str
    observed_at: str
    quality: str
    summary: str
    data_mode: DataMode


class DeploymentRecord(BaseModel):
    evidence_id: str
    version: str
    deployed_at: str
    author: str
    summary: str
    data_mode: DataMode


class HistoricalIncident(BaseModel):
    evidence_id: str
    incident_id: str
    root_cause: str
    summary: str
    data_mode: DataMode


class SopDocument(BaseModel):
    evidence_id: str
    sop_id: str
    title: str
    actions: list[str]
    data_mode: DataMode


class RetrievedContext(BaseModel):
    service: ServiceContext | None = None
    metrics: list[MetricSnapshot] = Field(default_factory=list)
    deployments: list[DeploymentRecord] = Field(default_factory=list)
    historical_incidents: list[HistoricalIncident] = Field(default_factory=list)
    sop_documents: list[SopDocument] = Field(default_factory=list)
    unknowns: list[str] = Field(default_factory=list)


class CandidateRootCause(BaseModel):
    cause: str
    evidence_ids: list[str]
    confidence: float = Field(ge=0.0, le=1.0)
    reasoning_summary: str


class RecommendedAction(BaseModel):
    action: str
    evidence_ids: list[str] = Field(default_factory=list)


class DiagnosisResult(BaseModel):
    candidate_root_causes: list[CandidateRootCause] = Field(default_factory=list)
    recommended_actions: list[RecommendedAction] = Field(default_factory=list)
    sop_refs: list[str] = Field(default_factory=list)
    unknowns: list[str] = Field(default_factory=list)
    abstentions: list[str] = Field(default_factory=list)


class RoutingRecommendation(BaseModel):
    recommended_team: str | None
    recommended_actions: list[RecommendedAction]
    escalation: str | None = None
    sop_refs: list[str] = Field(default_factory=list)


class FinalReport(BaseModel):
    title: str
    summary: str
    facts: list[str]
    derived_findings: list[str]
    assumptions: list[str]
    unknowns: list[str]
    recommendations: list[str]


class ReviewIssue(BaseModel):
    """ReviewerAgent 发现的具体问题。"""

    severity: Literal["critical", "warning", "info"]
    category: str
    description: str
    affected_fields: list[str] = Field(default_factory=list)
    evidence_ids: list[str] = Field(default_factory=list)


class ReviewResult(BaseModel):
    """ReviewerAgent 跨 Agent 证据一致性审查结果。"""

    review_status: Literal["consistent", "flagged", "abstain"]
    issues: list[ReviewIssue] = Field(default_factory=list)
    recommendation: str
    confidence: float = Field(ge=0.0, le=1.0)
    evidence_ids: list[str] = Field(default_factory=list)


class ReActToolDecision(BaseModel):
    """ReAct 循环中 LLM 的工具调用决策（伪工具调用）。"""

    action: Literal[
        "get_metrics", "get_deployments", "get_incidents", "get_sops", "finish"
    ]
    action_input: dict[str, Any] = Field(default_factory=dict)
    thought: str


class WorkflowStepTrace(BaseModel):
    step: str
    status: str
    started_at: str
    completed_at: str
    summary: str
    evidence_ids: list[str] = Field(default_factory=list)


class OpsTicketResult(BaseModel):
    affected_service: str | None
    candidate_root_causes: list[CandidateRootCause] = Field(default_factory=list)
    recommended_actions: list[RecommendedAction] = Field(default_factory=list)
    assigned_team: str | None
    sop_refs: list[str] = Field(default_factory=list)
    evidence_ids: list[str] = Field(default_factory=list)


class SupportTicketResult(BaseModel):
    request_type: str
    matched_articles: list[str] = Field(default_factory=list)
    reply_suggestions: list[str] = Field(default_factory=list)
    recommended_team: str | None
    escalation: str | None = None
    evidence_ids: list[str] = Field(default_factory=list)


class TicketProcessResponse(BaseModel):
    ticket_id: str
    run_id: str
    data_mode: DataMode
    classification: TicketClassification
    context: RetrievedContext
    diagnosis: DiagnosisResult
    routing: RoutingRecommendation
    review: ReviewResult | None = None
    report: FinalReport
    agent_trace: list[WorkflowStepTrace]
    evidence: list[Evidence]
    ops_result: OpsTicketResult | None = None
    support_result: SupportTicketResult | None = None
    similar_cases: list[dict[str, Any]] = Field(default_factory=list)
    notification: dict[str, Any] | None = None


class TicketProcessWsStartRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=2000)
    data_mode: DataMode = DataMode.MOCK
    desk_id: DeskId = DeskId.OPS

    @field_validator("text")
    @classmethod
    def strip_text(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("工单内容不能为空")
        return stripped


class TicketProcessWsStartMessage(BaseModel):
    type: Literal["start"]
    request: TicketProcessWsStartRequest


class TicketProcessWsCancelMessage(BaseModel):
    type: Literal["cancel"]
    reason: str | None = None


class TicketProcessWsEvent(BaseModel):
    type: str
    ticket_id: str
    run_id: str
    sequence: int
    timestamp: str = Field(default_factory=lambda: datetime.now(UTC).isoformat())


class TicketProcessWsStartedEvent(TicketProcessWsEvent):
    type: Literal["started"] = "started"


class TicketProcessWsAgentProgressEvent(TicketProcessWsEvent):
    type: Literal["agent_progress"] = "agent_progress"
    agent_name: str
    step: str
    status: str
    summary: str
    evidence_refs: list[EvidenceRef] = Field(default_factory=list)


class TicketProcessWsHeartbeatEvent(TicketProcessWsEvent):
    type: Literal["heartbeat"] = "heartbeat"


class TicketProcessWsCompletedEvent(TicketProcessWsEvent):
    type: Literal["completed"] = "completed"
    result: TicketProcessResponse


class TicketProcessWsCancelledEvent(TicketProcessWsEvent):
    type: Literal["cancelled"] = "cancelled"
    reason: str = "client_cancelled"


class TicketProcessWsErrorEvent(TicketProcessWsEvent):
    type: Literal["error"] = "error"
    error: dict[str, Any]

File: frontend/src/renderer/src/types/tickets.ts
export type DataMode = 'mock' | 'demo' | 'real'
export type DeskId = 'ops' | 'support'
export type RunStatus = 'completed' | 'failed' | 'cancelled' | 'pending'
export type TicketStatus = 'pending' | 'open' | 'in_progress' | 'resolved' | 'closed' | 'cancelled'
export type SupportReplyDraftStatus = 'draft' | 'approved' | 'sent' | 'discarded'

export interface LoginRequest {
  user_id: string
  password: string
}

export interface LoginResponse {
  token: string
  user_id: string
  name: string
  role: string
}

export interface TicketSubmitRequest {
  text: string
  desk_id: DeskId
}

export interface TicketSubmitResponse {
  ticket_id: string
  status: string
  created_at: string
  text: string
  desk_id: DeskId
  submitter: string
  assessed_priority?: string | null
  assessed_priority_reason?: string | null
}

export interface TicketHistorySummary {
  ticket_id: string
  desk_id: DeskId
  latest_run_id: string | null
  created_at: string
  updated_at: string
  data_mode: DataMode
  status: RunStatus
  ticket_status: TicketStatus
  submitter?: string | null
  assessed_priority?: string | null
  assessed_priority_reason?: string | null
  sla_deadline?: string | null
  claimed_by?: string | null
  claimed_at?: string | null
  assigned_team?: string | null
  resolution_summary?: string | null
  closed_at?: string | null
  summary?: string | null
  affected_service?: string | null
  priority?: string | null
  report_title?: string | null
}

export interface CatalogItem {
  evidence_id: string
  source_type: string
  source_id: string
  source_name: string
  retrieved_at?: string | null
  observed_at?: string | null
  desk_scope: DeskId
  id: string
  title: string
  category: string
  priority_hint: string
  affected_service: string
  description: string
  template_text: string
  quality: string
  data_mode: DataMode
  summary: string
}

export interface CatalogResponse {
  desk_id: DeskId
  data_mode: DataMode
  items: CatalogItem[]
}

export interface KnowledgeArticle {
  evidence_id: string
  source_type: string
  source_id: string
  source_name: string
  retrieved_at?: string | null
  observed_at?: string | null
  desk_scope: DeskId
  service: string
  article_id: string
  title: string
  actions: string[]
  quality: string
  data_mode: DataMode
  summary: string
  trace_uri?: string | null
  quality_reason?: string | null
}

export interface KnowledgeResponse {
  desk_id: DeskId
  data_mode: DataMode
  items: KnowledgeArticle[]
}

export interface TicketProcessRequest {
  text: string
  desk_id: DeskId
}

export interface EvidenceRef {
  evidence_id: string
}

export interface Evidence {
  evidence_id: string
  source_type: string
  source_id: string
  source_name: string
  observed_at?: string | null
  retrieved_at?: string | null
  service?: string | null
  metric_name?: string | null
  value?: unknown
  unit?: string | null
  quality: string
  data_mode: DataMode
  confidence?: number | null
  summary: string
  trace_uri?: string | null
  freshness?: string | null
  quality_reason?: string | null
  producer?: string | null
  run_id?: string | null
}

export interface TicketClassification {
  category: string
  summary: string
  affected_service: string | null
  symptoms: string[]
  priority: string
  priority_reason: string
  extracted_metrics: Record<string, unknown>
  evidence_ids: string[]
}

export interface ServiceContext {
  service_id: string
  name: string
  display_name: string
  aliases: string[]
  owner_team: string
  criticality: string
  dependencies: string[]
  data_mode: DataMode
}

export interface MetricSnapshot {
  evidence_id: string
  metric_name: string
  value: unknown
  unit: string
  observed_at: string
  quality: string
  summary: string
  data_mode: DataMode
}

export interface DeploymentRecord {
  evidence_id: string
  version: string
  deployed_at: string
  author: string
  summary: string
  data_mode: DataMode
}

export interface HistoricalIncident {
  evidence_id: string
  incident_id: string
  root_cause: string
  summary: string
  data_mode: DataMode
}

export interface SopDocument {
  evidence_id: string
  sop_id: string
  title: string
  actions: string[]
  data_mode: DataMode
}

export interface RetrievedContext {
  service: ServiceContext | null
  metrics: MetricSnapshot[]
  deployments: DeploymentRecord[]
  historical_incidents: HistoricalIncident[]
  sop_documents: SopDocument[]
  unknowns: string[]
}

export interface CandidateRootCause {
  cause: string
  evidence_ids: string[]
  confidence: number
  reasoning_summary: string
}

export interface DiagnosisResult {
  candidate_root_causes: CandidateRootCause[]
  unknowns: string[]
  abstentions: string[]
}

export interface RecommendedAction {
  action: string
  evidence_ids: string[]
}

export interface RoutingRecommendation {
  recommended_team: string | null
  recommended_actions: RecommendedAction[]
  escalation: string
  sop_refs: string[]
}

export interface FinalReport {
  title: string
  summary: string
  facts: string[]
  derived_findings: string[]
  assumptions: string[]
  unknowns: string[]
  recommendations: string[]
}

export interface WorkflowStepTrace {
  step: string
  status: string
  started_at: string
  completed_at: string
  summary: string
  evidence_ids: string[]
}

export interface OpsTicketResult {
  affected_service: string | null
  candidate_root_causes: CandidateRootCause[]
  recommended_actions: RecommendedAction[]
  assigned_team: string | null
  sop_refs: string[]
  evidence_ids: string[]
}

export interface SupportTicketResult {
  request_type: string
  matched_articles: string[]
  reply_suggestions: string[]
  recommended_team: string | null
  escalation: string
  evidence_ids: string[]
}

export interface TicketProcessResponse {
  ticket_id: string
  run_id: string
  data_mode: DataMode
  classification: TicketClassification
  context: RetrievedContext
  diagnosis: DiagnosisResult
  routing: RoutingRecommendation
  report: FinalReport
  agent_trace: WorkflowStepTrace[]
  evidence: Evidence[]
  review?: ReviewResult | null
  ops_result?: OpsTicketResult | null
  support_result?: SupportTicketResult | null
  notification?: Record<string, unknown> | null
}

export interface ReviewIssue {
  severity: 'critical' | 'warning' | 'info'
  category: string
  description: string
  affected_fields: string[]
  evidence_ids: string[]
}

export interface ReviewResult {
  review_status: 'consistent' | 'flagged' | 'abstain'
  issues: ReviewIssue[]
  recommendation: string
  confidence: number
  evidence_ids: string[]
}

export interface ReActStep {
  step_index: number
  decision_summary: string
  action: string
  action_input_summary: string
  observation_summary: string
  evidence_ids: string[]
}

export interface AgentTaskError {
  code: string
  message: string
  details: Record<string, unknown>
}

export interface StoredAgentRun {
  sequence: number
  task_id: string
  agent_name: string
  step: string
  status: string
  route_decision: unknown | null
  observations: string[]
  react_steps?: ReActStep[]
  evidence_ids: string[]
  error: AgentTaskError | null
  started_at: string
  completed_at?: string | null
}

export interface TicketLifecycleUpdateRequest {
  ticket_status?: TicketStatus
  resolution_summary?: string | null
  root_cause?: string | null
  fix_action?: string | null
  verification?: string | null
}

export interface SupportReplyDraftUpdateRequest {
  reply_text: string
  report_summary?: string | null
  evidence_ids: string[]
  status?: SupportReplyDraftStatus
  editor?: string | null
}

export interface SupportReplyDraftResponse {
  draft_id: string
  ticket_id: string
  run_id: string
  source: string
  reply_text: string
  report_summary?: string | null
  evidence_ids: string[]
  status: SupportReplyDraftStatus
  editor?: string | null
  created_at: string
  updated_at: string
  approved_at?: string | null
  sent_at?: string | null
}

export interface TicketHistoryListResponse {
  items: TicketHistorySummary[]
  limit: number
  offset: number
  total: number
}

export interface StoredRunDetail {
  run_id: string
  status: RunStatus
  data_mode: DataMode
  route_mode: string
  started_at: string
  completed_at: string
  response: TicketProcessResponse | null
  error: AgentTaskError | null
  agent_runs: StoredAgentRun[]
  supervisor_decisions: unknown[]
  evidence: Evidence[]
}

export interface TicketHistoryDetailResponse {
  ticket_id: string
  desk_id: DeskId
  input_text: string
  data_mode: DataMode
  ticket_status: TicketStatus
  submitter?: string | null
  assigned_team?: string | null
  resolution_summary?: string | null
  root_cause?: string | null
  fix_action?: string | null
  verification?: string | null
  closed_at?: string | null
  created_at: string
  updated_at: string
  support_reply_draft?: SupportReplyDraftResponse | null
  latest_run: StoredRunDetail | null
}

export interface ApiErrorPayload {
  error: {
    code: string
    message: string
    details: Record<string, unknown>
  }
}

export interface TicketProcessWsStartedEvent {
  type: 'started'
  ticket_id: string
  run_id: string
  sequence: number
  timestamp: string
}

export interface TicketProcessWsAgentProgressEvent {
  type: 'agent_progress'
  ticket_id: string
  run_id: string
  sequence: number
  timestamp: string
  agent_name: string
  step: string
  status: string
  summary: string
  evidence_refs: EvidenceRef[]
}

export interface TicketProcessWsCompletedEvent {
  type: 'completed'
  ticket_id: string
  run_id: string
  sequence: number
  timestamp: string
  result: TicketProcessResponse
}

export interface TicketProcessWsCancelledEvent {
  type: 'cancelled'
  ticket_id: string
  run_id: string
  sequence: number
  timestamp: string
  reason: string
}

export interface TicketProcessWsErrorEvent {
  type: 'error'
  ticket_id: string
  run_id: string
  sequence: number
  timestamp: string
  error: ApiErrorPayload['error']
}

export type TicketProcessWsEvent =
  | TicketProcessWsStartedEvent
  | TicketProcessWsAgentProgressEvent
  | TicketProcessWsCompletedEvent
  | TicketProcessWsCancelledEvent
  | TicketProcessWsErrorEvent
